### ISSUE TITLE
Python

### ISSUE THREAD

**Issue by DSP137** [openconnectome/Statistical-Connectomics-Class#231]
So, not a paper related question, but I am planning on downloading and learning Python (or at least the basics) over the summer. There are two versions out (2.x and 3.x). If I am going to want to eventually do statistics on graphs and networks using code written in Python, which version of Python would be better to get? Also, are there any recommendations on what would be good to know (in particular)?

**Comment by gkiar** [comment]
Though the world will eventually migrate to Python 3.x, most packages that
are made are still made for/compatible with Python 2.7.x, in my experience.
Pretty well everyone in our lab uses Python 2.7.x.

In terms of what to know? I'm not sure I understand fully, but if you're
looking for tools/libraries then I'd make sure you're aware of: scipy,
numpy, igraph, networkx. I'm sure there are others as well, but they
currently escape me. :) If you have more specific function requests feel
free to let me know, but those should cover a vast majority of what you
want to do on the math side of things. Is that what you were looking for?

--
Greg Kiar <gkiar07@gmail.com>

**Comment by ElanHR** [comment]
If you're interested in using python for statistics and care about performance (or want to use some really fast prepackaged machine-learning algorithms) I'd recommend checking out [Theano](http://deeplearning.net/software/theano/). The learning curve is a bit steep but I was pretty much new to python as well at the beginning of the semester and I think at this point I'm fairly proficient.

**Comment by whock** [comment]
I agree with what's been said. When thinking about making the leap to a new version of something I think of it as two competing deadlines: how long will the old version be supported versus how long until all the stuff I need is supported in the new version? In my mind right now for Python I'm leaning towards the old versions supported longer. So I'm using 2.7.x and will do so for the foreseeable future.

**Comment by mrjiaruiwang** [comment]
I had the unfortunate pleasure of learning Python 3 and learned quickly that the community is pretty much evenly divided between those still hanging onto 2.7 and those who have moved on to 3. The differences are significant enough that you cannot easily switch between the two. However, my recommendation would still be to learn 3 just because that is what is being taught now and the new generation will eventually replace the old, but for now the pain is real. Macs come with 2.7 installed by default. I would look into parallel programming too as many of the algorithms we explored in this class tends to be computational hard.

**Comment by imichaelnorris** [comment]
I'm going to disagree with @gkiar here.  Based on http://py3readiness.org/ I think Python 3 is ready for primetime.  2 or 3 years ago all packages weren't ready but today the only things that aren't ready for Python 3 are really esoteric web development libraries.

I don't think the differences are too large between Python 2 and 3 syntactically.

**Comment by jovo** [comment]
@gkiar - which packages do we use that don't like py3?

On Sun, Apr 26, 2015 at 4:05 PM, Michael Norris <notifications@github.com>
wrote:

>  I'm going to disagree with @gkiar <https://github.com/gkiar> here. Based
> on http://py3readiness.org/ I think Python 3 is ready for primetime. 2 or
> 3 years ago all packages weren't ready but today the only things that
> aren't ready for Python 3 are really esoteric web development libraries.
>
> I don't think the differences are too large between Python 2 and 3
> syntactically.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openconnectome/Statistical-Connectomics-Class/issues/231#issuecomment-96429370>
> .
>



-- 
the glass is all full:  half water, half air.
openconnecto.me, jovo.me, office hours
<https://www.google.com/calendar/embed?src=e2ktu4lrgul8anp8hclrcminp8%40group.calendar.google.com&ctz=America/New_York>

**Comment by gkiar** [comment]
igraph is the big one I'm aware of, since we use an older version. Also
django. I can't speak for everyone else, but I also use 2.7 because it is
what the community mostly uses - more people use it so more support/bug
fixes/googleable answers... \to easier development.

--
Greg Kiar <gkiar07@gmail.com>

**Comment by mrjiaruiwang** [comment]
You can go from either 3 to 2 or 2 to 3 using the functions 3to2 and 2to3 (https://docs.python.org/2/library/2to3.html). This won't work perfectly as it only deals with syntax, but it should be a good start. I would say learning 3 and then dealing with porting would be easier than learning 2 and then watching the world slowly shift to 3, then having to learn the differences.

**Comment by willgray** [comment]
I still don't have a clear view of when python 3.x adoption...but lots of existing algorithms are in 2.7 still. 

Other packages that we use a lot:

scikit-learn
scikit-image

Anaconda (better?) and Enthought Canopy are both free distributions with lots of packages and some amount of integration (e.g. Canopy has a functional IDE that looks kinda sorta like matlab)
